### ISSUE TITLE
Container.withError

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#8402]
## Problem

When you have a beautiful chain of 20 function calls, no error checking end-to-end... except the one in the middle, then you're forced to break your chain, plumb error checking all the way up... So much beauty ruined.

## Solution

Allow me to attach an error to the container, so that it is lazily returned to the caller, and I don't have to deal with error passing in my code.  Example:

```golang
func Beautiful(ctx context.Context) *dagger.Container {
  return dag.Container().
    Foo().
    Bar().
    Baz().
    With(func(c *dagger.Container) *dagger.Container {
      contents, err := c.
        WithExec([]string{"check", "the", "thing"}).
        File("the-thing-output.txt").
        Contents(ctx)
      if err != nil {
        return c.WithError(err)
      }
      if contents != "expected-thing" {
        return c.WithErrorf(fmt.Errorf("The thing wasn't the expected shape! Failed to check the thing"))
      }
      return c
    }).
    Bla().
    Bla().
    ChefsKiss()
}
```

**Comment by jedevc** [comment]
Agreed, I previously considered whether we could *instead* have `WithErr` (naming tbd), which takes a `func (*dagger.Container) (*dagger.Container, error)`. I'd rather we don't *explicitly* embed errors into containers, but I'm struggling to explain why I think that pattern is less desirable.

Regardless of the interface, the error would then get surfaced in the next function in the chain which returns an `error`, and if one is not present, it would be exposed at the top level of the module.

I like the idea :tada:

---

After an error is returned, do all the later steps get skipped - including future `With`s? I would assume so. Kinda neatly, we're just building monadic do-notation in go with this :laughing:
